### ISSUE TITLE
Question: How to build some task scheduler like Unix?

### ISSUE THREAD

**Issue by master-q** [boogie-org/corral#33]
I wrote following script:

```
var ready: int;
var gval: int;
var tick: int;

procedure {:entrypoint} main()
{
  call pre_init();
  async call dotick();
  async call thread();
  call wait(5);
  call set_gval();
  call wait(5);
  call wakeup_ready();
}

procedure set_gval()
{
  gval := 10;
}

procedure wakeup_ready()
{
  ready := 1;
}

procedure pre_init()
{
  ready := 0;
  gval := 0;
  tick := 0;
}

procedure thread()
{
  call wait_ready(15);
  assert gval == 10; // Why?
}

procedure dotick()
{
  while (true)
  {
    call ticktack();
  }
}

procedure ticktack()
{
  yield;
  tick := tick + 1;
  yield;
}

procedure wait(w: int)
{
  var until: int;
  until := tick + w;
  while (tick < until) {
    yield;
  }
}

procedure wait_uncertain(w: int)
{
  var h: int;
  var until: int;
  yield;
  havoc h;
  assume (0 <= h && h <= w);
  until := tick + h;
  while (tick < until) {
    yield;
  }
}

procedure wait_ready(w: int)
{
  var start: int;
  var until: int;
  until := tick + w;
  yield;
  while (ready == 0 && tick < until)
  {
    yield;
  }
  yield;
}
```

This script has three tasks. They run as following:

First: "Main" task
1. fork "Tick" task
2. fork "Thread" task
3. wait 5 second
4. write `10` into `gval`
5. wait 5 second
6. write `1` into `ready` for wakeup "Tick" task

Second: "Tick" task
1. `while (true) { tick := tick + 1 }`

Third: "Thread" task
1. wait 15 second, but wakeup if `ready == 1`
2. `assert gval == 10`

However, this script does not make sense, because "Tick" task does not encourage the other tasks to run.

How to build such task scheduler like Unix on Corral?

**Comment by akashlal** [comment]
I am not sure what you mean by "encourage other tasks to run"? It seems that you want to assign a special meaning to a "tick" but I don't understand.

**Comment by master-q** [comment]
Sorry for my poor explanation.

My motivation is to verify a script language by using Corral. The language has task like pthread. Tasks can share global value. Also, the language has `wait` function which is wait until specified time like sleep(3) in C language. If waiting task receives trigger from another task, then the `wait` function immediately.

> It seems that you want to assign a special meaning to a "tick" but I don't understand.

The "tick" is meaning the system global clock. In this Corral script, the `wait` function waits until that the "tick" reaches specified time. The trick is to emulate actual `wait` function in Corral.

> I am not sure what you mean by "encourage other tasks to run"?

My question is "how to build such `wait` function in Corral world"? In my script, `wait` function will not be waked up, because the `dotick` function does not emulate the system global clock, perfectly.

I hope this script runs as following:

1. Main task: fork "Tick" task
2. Main task: fork "Thread" task
3. Main task: wait 5 second
4. Thread task: wait 15 second
5. Main task: write 10 into `gval`
6. Main task: wait 5 second
7. Main task: wake up Thread task
8. Thread task: wake up from wait
9. Thread task: Do `assert gval == 10`

**Comment by akashlal** [comment]
Do you want tick to run synchronously with the other threads? That is, if tick is X and some thread executes wait(5) then the thread returns immediately when tick reaches X+5? In the current semantics of the language, you only get the guarantee that the thread will be woken up at an arbitrary time after X+5. 

In your current program, the following is a valid behavior:
1. Main task forks "Tick" task
2. Main task forms "Thread" task
3. Main executes wait(5)
4. Thread executes wait_ready(15)
5. Tick runs for 15 iterations, sets tick to 15.
6. Thread's wait finishes, it executes the assert and fails.

That is, your program does not force the main thread to run as soon as tick reaches 5. In order to do this, you must use additional synchronization. If some thread executes wait(5), then it also inserts a block for the Tick task to stop when it increments tick by 5. The Tick task will stop at tick+5. Then wait(5) returns and unblocks Tick.

Makes sense? It seems that modeling synchronous computation is a bit difficult.

**Comment by master-q** [comment]
> Do you want tick to run synchronously with the other threads?

Yes.

> That is, if tick is X and some thread executes wait(5) then the thread returns immediately when tick reaches X+5?

Yes.

> In the current semantics of the language, you only get the guarantee that the thread will be woken up at an arbitrary time after X+5.

Umm... Then you mean I can't wake up the other thread.

> That is, your program does not force the main thread to run as soon as tick reaches 5.

I understand it, now.

> In order to do this, you must use additional synchronization.
> If some thread executes wait(5), then it also inserts a block for the Tick task to stop when it increments tick by 5.
> The Tick task will stop at tick+5. Then wait(5) returns and unblocks Tick.

O.K. I'll try to do such trick.

> Makes sense? It seems that modeling synchronous computation is a bit difficult.

"Difficult" does not mean "impossible"?
